clients/githubrepo/branches.go

```diff
@@ -534,37 +557,49 @@ func translateRequiredStatusRepoRule(base *clients.BranchProtectionRule, rule *r
 	}
 }
 
+// Merge strategy:
+//   - if both are nil, keep it nil
+//   - if any of them is not nil, keep the most restrictive one
 func mergeBranchProtectionRules(base, translated *clients.BranchProtectionRule) {
-	if base.AllowDeletions == nil || translated.AllowDeletions != nil && !*translated.AllowDeletions {
+	if base.AllowDeletions == nil || (translated.AllowDeletions != nil && !*translated.AllowDeletions) {
 		base.AllowDeletions = translated.AllowDeletions
 	}
-	if base.AllowForcePushes == nil || translated.AllowForcePushes != nil && !*translated.AllowForcePushes {
+	if base.AllowForcePushes == nil || (translated.AllowForcePushes != nil && !*translated.AllowForcePushes) {
 		base.AllowForcePushes = translated.AllowForcePushes
 	}
-	if base.EnforceAdmins == nil || translated.EnforceAdmins != nil && !*translated.EnforceAdmins {
+	if base.EnforceAdmins == nil || (translated.EnforceAdmins != nil && !*translated.EnforceAdmins) {
 		// this is an over simplification to get preliminary support for repo rules merged.
 		// A more complete approach would process all rules without bypass actors first,
 		// then process those with bypass actors. If no settings improve (due to rule layering),
 		// then we can ignore the bypass actors.
 		// https://github.com/ossf/scorecard/issues/3480
 		base.EnforceAdmins = translated.EnforceAdmins
 	}
-	if base.RequireLastPushApproval == nil || readBoolPtr(translated.RequireLastPushApproval) {
-		base.RequireLastPushApproval = translated.RequireLastPushApproval
+	if base.RequireLastPushApproval != nil || translated.RequireLastPushApproval != nil {
+		base.RequireLastPushApproval = initializedBoolRef(
+			readBoolPtr(base.RequireLastPushApproval) || readBoolPtr(translated.RequireLastPushApproval),
+		)
 	}
-	if base.RequireLinearHistory == nil || readBoolPtr(translated.RequireLinearHistory) {
-		base.RequireLinearHistory = translated.RequireLinearHistory
+	if base.RequireLinearHistory != nil || translated.RequireLinearHistory != nil {
+		base.RequireLinearHistory = initializedBoolRef(
+			readBoolPtr(base.RequireLinearHistory) || readBoolPtr(translated.RequireLinearHistory),
+		)
 	}
-	mergePullRequestReviews(&base.RequiredPullRequestReviews, &translated.RequiredPullRequestReviews)
 	mergeCheckRules(&base.CheckRules, &translated.CheckRules)
+	base.RequiredPullRequestReviews = mergePullRequestReviews(
```

Comment:
I think the old code handled the layering fine and avoided all the extra allocations

Can we keep the old format and just change the type of base to a double pointer i think?